README.md
<a href="https://www.gotoiot.com/">
    <img src="doc/gotoiot-logo.png" alt="logo" title="Goto IoT" align="right" width="60" height="60" />
</a>

Web App Full Stack Base
=======================

*Ayudaría mucho si apoyaras este proyecto con una ⭐ en Github!*

Este proyecto es una aplicación web fullstack que se ejecuta sobre el ecosistema `Docker`. Está compuesta por un compilador de `TypeScript` que te permite utilizar este superset de JavaScript para poder programar un `cliente web`. También tiene un servicio en `NodeJS` que te permite ejecutar código en backend y al mismo tiempo disponibilizar el código del cliente web para interactar con el servicio. Además tiene una `base de datos` MySQL que puede interactuar con el backend para guardar y consultar datos, y de manera adicional trae un `administrador` de base de datos para poder administrar la base en caso que lo necesites.

La aplicación IoT de base que viene con este proyecto se encarga de crear una tabla llamada `Devices` en la base de datos, y la idea es que vos puedas desarrollar el código de backend y frontend que te permita controlar desde el navegador el estado de los devices de un hogar inteligente - *como pueden ser luces, TVs, ventiladores, persianas, enchufes y otros* - y almacenar los estados de cada uno en la base de datos. 

Realizando estas tareas vas a a tener una aplicación fullstack IoT del mundo real que utiliza tecnologías actuales en la que un backend es capaz de interactuar con una DB para cumplir con las peticiones de control que se le mandan desde el cliente web.

En esta imagen podés ver una posible implementación del cliente web que controla los artefactos del hogar.

![architecture](doc/webapp-example-1.png)

## Comenzando 🚀

Esta sección es una guía con los pasos escenciales para que puedas poner en marcha la aplicación.

<details><summary><b>Mira los pasos necesarios</b></summary><br>

### Instalar las dependencias

Para correr este proyecto es necesario que instales `Docker` y `Docker Compose`. 

En [este artículo](https://www.gotoiot.com/pages/articles/docker_installation_linux/) publicado en nuestra web están los detalles para instalar Docker y Docker Compose en una máquina Linux. Si querés instalar ambas herramientas en una Raspberry Pi podés seguir [este artículo](https://www.gotoiot.com/pages/articles/rpi_docker_installation) de nuestra web que te muestra todos los pasos necesarios.

En caso que quieras instalar las herramientas en otra plataforma o tengas algún incoveniente, podes leer la documentación oficial de [Docker](https://docs.docker.com/get-docker/) y también la de [Docker Compose](https://docs.docker.com/compose/install/).

Continua con la descarga del código cuando tengas las dependencias instaladas y funcionando.

### Descargar el código

Para descargar el código, usar el comando siguiente:

```
git clone https://github.com/rlosada/app-fullstack-base.git
```
### Ejecutar la aplicación

Para ejecutar la aplicación tenes que correr el comando `docker-compose up` desde la raíz del proyecto. Este comando va a descargar las imágenes de Docker de node, de typescript, de la base datos y del admin de la DB, y luego ponerlas en funcionamiento. 

Para acceder al cliente web ingresa a a la URL [http://localhost:8000/](http://localhost:8000/) y para acceder al admin de la DB accedé a [localhost:8001/](http://localhost:8001/). 

Si pudiste acceder al cliente web y al administrador significa que la aplicación se encuentra corriendo bien. 

> Si te aparece un error la primera vez que corres la app, deteńe el proceso y volvé a iniciarla. Esto es debido a que el backend espera que la DB esté creada al iniciar, y en la primera ejecución puede no alcanzar a crearse. A partir de la segunda vez el problema queda solucionado.



</details>

Continuá explorando el proyecto una vez que lo tengas funcionando.

## Configuraciones de funcionamiento 🔩

Al crearse la aplicación se ejecutan los contenedores de Docker de cada servicio, se crea la base de datos y sus tablas. A continuación podés encontrar info si querés cambiar la estructura de la DB o bien sus configuraciones de acceso.

<details><summary><b>Lee cómo configurar la aplicación</b></summary><br>

### Configuración de la DB

Como ya comprobaste, para acceder PHPMyAdmin tenés que ingresar en la URL [localhost:8001/](http://localhost:8001/). En el login del administrador, el usuario para acceder a la db es `root` y contraseña es la variable `MYSQL_ROOT_PASSWORD` del archivo `docker-compose.yml`.

Para el caso del servicio de NodeJS que se comunica con la DB fijate que en el archivo `src/backend/mysql-connector.js` están los datos de acceso para ingresar a la base.

Si quisieras cambiar la contraseña, puertos, hostname u otras configuraciones de la DB deberías primero modificar el servicio de la DB en el archivo `docker-compose.yml` y luego actualizar las configuraciones para acceder desde PHPMyAdmin y el servicio de NodeJS.

### Estructura de la DB

Al iniciar el servicio de la base de datos, si esta no está creada toma el archivo que se encuentra en `db/dumps/smart_home.sql` para crear la base de datos automáticamente.

En ese archivo está la configuración de la tabla `Devices` y otras configuraciones más. Si quisieras cambiar algunas configuraciones deberías modificar este archivo y crear nuevamente la base de datos para que se tomen en cuenta los cambios.

Tené en cuenta que la base de datos se crea con permisos de superusuario por lo que no podrías borrar el directorio con tu usuario de sistema, para eso debés hacerlo con permisos de administrador. En ese caso podés ejecutar el comando `sudo rm -r db/data` para borrar el directorio completo.

### Lanzar la aplicacion

Una vez descargada la aplicacion en el host donde la misma será ejecutada es preciso realizar los siguientes pasos:

1) Instalar dependencias:

```
cd app-fullstack-base/src/backend
npm install
```
Al proyecto orinal se le agrego una dependencia adicional, status-code-enum. 

2) Editar la URL del backend a utilizar. El valor asignado debera ser la ip del host donde se corra la aplicacion. Para ello abrir el archivo app-fullstack-base/src/frontend/common.ts
y editar la linea 32.

```
const BACKEND_URL = 'http://192.168.0.2:8000';
```

En este caso la ip 192.168.0.2 es la ip donde se va a hostear la aplicacion.

3) Finalemente lanzar la aplicacion (los pasos 1 y 2 solo deberan hacerse una vez):

```
cd app-fullstack-base
docker-compose up
```

</details>


## Detalles principales 🔍

En esta sección vas a encontrar las características más relevantes del proyecto.

<details><summary><b>Mira los detalles más importantes de la aplicación</b></summary><br>
<br>

### Arquitectura de la aplicación

Como ya pudiste ver, la aplicación se ejecuta sobre el ecosistema Docker, y en esta imagen podés ver el diagrama de arquitectura.

![architecture](doc/architecture.png)

### El cliente web

El cliente web es una Single Page Application que se comunica con el servicio en NodeJS mediante JSON a través de requests HTTP. Puede consultar el estado de dispositivos en la base de datos (por medio del servicio en NodeJS) y también cambiar el estado de los mismos. Los estilos del código están basados en **Material Design**.

### El servicio web

El servicio en **NodeJS** posee distintos endpoints para comunicarse con el cliente web mediante requests HTTP enviando **JSON** en cada transacción. Procesando estos requests es capaz de comunicarse con la base de datos para consultar y controlar el estado de los dispositivos, y devolverle una respuesta al cliente web también en formato JSON. Así mismo el servicio es capaz de servir el código del cliente web.

### La base de datos

La base de datos se comunica con el servicio de NodeJS y permite almacenar el estado de los dispositivos en la tabla **Devices**. Ejecuta un motor **MySQL versión 5.7** y permite que la comunicación con sus clientes pueda realizarse usando usuario y contraseña en texto plano. En versiones posteriores es necesario brindar claves de acceso, por este motivo la versión 5.7 es bastante utilizada para fases de desarrollo.

### El administrador de la DB

Para esta aplicación se usa **PHPMyAdmin**, que es un administrador de base de datos web muy utilizado y que podés utilizar en caso que quieras realizar operaciones con la base, como crear tablas, modificar columnas, hacer consultas y otras cosas más.

### El compilador de TypeScript

**TypeScript** es un lenguaje de programación libre y de código abierto desarrollado y mantenido por Microsoft. Es un superconjunto de JavaScript, que esencialmente añade tipos estáticos y objetos basados en clases. Para esta aplicación se usa un compilador de TypeScript basado en una imagen de [Harmish](https://hub.docker.com/r/harmish) en Dockerhub, y está configurado para monitorear en tiempo real los cambios que se realizan sobre el directorio **src/frontend/ts** y automáticamente generar código compilado a JavaScript en el directorio  **src/frontend/js**. Los mensajes del compilador aparecen automáticamente en la terminal al ejecutar el comando **docker-compose up**.

### Ejecución de servicios

Los servicios de la aplicación se ejecutan sobre **contenedores de Docker**, así se pueden desplegar de igual manera en diferentes plataformas. Los detalles sobre cómo funcionan los servicios los podés ver directamente en el archivo **docker-compose.yml**.

### Organización del proyecto

En la siguiente ilustración podés ver cómo está organizado el proyecto para que tengas en claro qué cosas hay en cada lugar.

```sh
├── db                          # directorio de la DB
│   ├── data                    # estructura y datos de la DB
│   └── dumps                   # directorio de estructuras de la DB
│       └── smart_home.sql      # estructura con la base de datos "smart_home"
├── doc                         # documentacion general del proyecto
└── src                         # directorio codigo fuente
│   ├── backend                 # directorio para el backend de la aplicacion
│   │   ├── index.js            # codigo principal del backend
│   │   ├── mysql-connector.js  # codigo de conexion a la base de datos
│   │   ├── package.json        # configuracion de proyecto NodeJS
│   │   └── package-lock.json   # configuracion de proyecto NodeJS
│   │   └── routes              # codigo que procesa las requests HTTP soportadas
│   │   └── common.js           # codigo y definiciones comunes 
│   │   └── logger.js           # codigo utilizado para generar los mensajes de log
│   └── frontend                # directorio para el frontend de la aplicacion
│       ├── js                  # codigo javascript que se compila automáticamente
│       ├── static              # donde alojan archivos de estilos, imagenes, fuentes, etc.
│       ├── ts                  # donde se encuentra el codigo TypeScript a desarrollar
│       └── index.html          # archivo principal del cliente HTML
├── docker-compose.yml          # archivo donde se aloja la configuracion completa
├── README.md                   # este archivo
├── CHANGELOG.md                # archivo para guardar los cambios del proyecto
├── LICENSE.md                  # licencia del proyecto
```

> No olvides ir poniendo tus cambios en el archivo `CHANGELOG.md` a medida que avanzas en el proyecto.

</details>

## Detalles de implementación 💻

En esta sección podés ver los detalles específicos de funcionamiento del código y que son los siguientes.

<details><summary><b>Mira los detalles de implementación</b></summary><br>

### Agregar un dispositivo

Para agregar un nuevo dispositivo en la pantalla principal hacer click en el boton '+Agregar', esto presentara
un formulario que hay que completar y luego para confirmar la creacion hacer click en el boton 'Crear'. Esto
lanzara el proceso de creación del dispositivo y luego la pagina se refrescara volviendo a mostrar el listado
completo de dispositivos donde si todo salio bien aparecera listado el nuevo dispositivo. En caso de que se desee
cancelar la operacion en vez de hacer click en 'Crear' hacerlo en 'Cancelar' lo que volvera a mostrar la lista
de dispositivos existente.

### Eliminar un dispositivo

Para eliminar un dispositivo existente solo hacer click en el boton 'Eliminar' ubicado a la derecha del dispositivo
que se desea eliminar. No existe confirmacion con lo cual una vez hecho el click la orden se ejecuta  y el dispositivo
se elimina. Luego la pantalla se refrescara mostrando los dispositivos restantes en la base.

### Editar un dispositivo

Para editar un dispositivo existente en la pantalla principal hacer click en el boton 'Editar' ubicado a la derecha del dispositivo
que se desea editar. Esto mostrara un formulario ya inicializado con los datos del dispositivo seleccionado. Realizar los
cambios necesarios y luego hacer click en 'Aplicar' para confirmar los cambio o 'Cancelar' para cancelar. En cualquiera de los dos
casos luego se vuelve automaticamente a la pantalla principal donde se listaran todos los dispositivos existentes.


### Frontend

El Frontend consta de dos pantalla. La primera es la que se carga inicialmente y muestra la lista de dispositivos recuperados del 
backend. Cada uno de los elementos de esta lista presenta dos botones para interactuar. 'Editar' que permite editar los valores del
dispositivo y 'Eliminar' que permite eliminar el dispositivo del backend.

La segunda pantalla es la usada para completar los datos de un dispositivo a crear o bien para editar los datos de un dispositivo existente.
En esta pantalla apareceran botones para 'Crear|Aplicar' o 'Cancelar'. En el primer caso se crea el dispositivo o se editan sus valores en
el backend y el segundo caso la operacion es cancelada y se vuelve a la lista de dispositivos.

### Backend

El backend implementa 5 endpoints que permiten recuperar los dispositivos de la base de datos, insertar nuevos dispositivos, actualizar dispositivos
existente y eliminar dispositivos existentes.

<details><summary><b>Ver los endpoints disponibles</b></summary><br>

Completá todos los endpoints del backend con los metodos disponibles, los headers y body que recibe, lo que devuelve, ejemplos, etc.

1) Endpoint : GET /devices

Recupera la lista de dispositivos de la base de datos. El formato devuelto es un string que contiene un json. En dicho json habra un arreglo de dispositivos. Si no hay entradas
en la tabla se devuelve un arreglo vacio. A continuacion un ejemplo:

```
[
    {"id": 8, "name":"Lampara", "description":"Lampara en el living", "state":1, "type":2},
    {"id":10, "name":"Lampara101","description":"esta es la lampara que esta ubicada en la habitacion que contiene los floreros","state":1,"type":2},
    {"id":11, "name":"aire acondicionado","description":"este es el aire acondicionado del living","state":0,"type":1}
]
```

2) Endpoint : GET /devices/:id

Es un caso especial del anterior, aqui se devuevle un arreglo que solo contiene el elemento cuyo id coincide con el de la tabla Devices de la base de datos. Si no hay concidencia
se devuelve un arreglo vacio.

```
[
    {"id": 8, "name":"Lampara", "description":"Lampara en el living", "state":1, "type":2},
]
```

3) Endpoint : DELETE /devices/:id

En este caso esta solicitud provoca que el backend elimine el dispositivo cuyo id coincide con el de la tabla Devices de la base. Si la eliminacion fue satisfactoria se devuelve un json que contiene el id eliminado. Si el dispositivo no existe, igualmente la respuesta es la misma. Si hay un error se devuelve un '500 Interal Server Error'.

```
{
    "id": "8"
}
```

4) Endpoint : POST /devices/

En este caso esta solicitud provoca que el backend cree un nuevo dispositivo en la tabla Devices. El content-type debera ser de tipo 'application/json' y el contenido del body debera
ser un objeto tipo device valido. A continuacion un ejemplo:

```
{
  "name" : "Televisor",
  "description" : "televisor ubicado en el comedor",
  "state" : 1,
  "type" : 0
}
```
Si el json esta mal formado o bien el content-type es incorrecto se devolvera un '404 Bad Request', si en cambio ocurre algun error interno '500 Interal Server Error'. Si todo funciona 
se devolvera un '2OO Ok' y un json con el id asigando al nuevo dispositivo en la base tal y como se muestra en el ejemplo a continuacion:

```
{
    "id": "8"
}
```

4) Endpoint : PATCH /devices/:id

En este caso esta solicitud provoca que el backend edite un dispositivo existente en la tabla Devices. El content-type debera ser de tipo 'application/json' y el contenido del body debera
ser un objeto tipo device valido. El id en la ruta identifica el dispositivo a editar. A continuacion un ejemplo:

```
{
  "name" : "Televisor",
  "description" : "televisor ubicado en el comedor",
  "state" : 1,
  "type" : 0
}
```
Si el json esta mal formado o bien el content-type es incorrecto se devolvera un '404 Bad Request', si en cambio ocurre algun error interno '500 Interal Server Error'. Si todo funciona 
se devolvera un '2OO Ok' y un json con el id del dispositivo editado en la base tal y como se muestra en el ejemplo a continuacion:

```
{
    "id": "8"
}
```

</details>

</details>


## Tecnologías utilizadas 🛠️

En esta sección podés ver las tecnologías más importantes utilizadas.

<details><summary><b>Mira la lista completa de tecnologías</b></summary><br>

* [Docker](https://www.docker.com/) - Ecosistema que permite la ejecución de contenedores de software.
* [Docker Compose](https://docs.docker.com/compose/) - Herramienta que permite administrar múltiples contenedores de Docker.
* [Node JS](https://nodejs.org/es/) - Motor de ejecución de código JavaScript en backend.
* [MySQL](https://www.mysql.com/) - Base de datos para consultar y almacenar datos.
* [PHPMyAdmin](https://www.phpmyadmin.net/) - Administrador web de base de datos.
* [Material Design](https://material.io/design) - Bibliotecas de estilo responsive para aplicaciones web.
* [TypeScript](https://www.typescriptlang.org/) - Superset de JavaScript tipado y con clases.

</details>

## Contribuir 🖇️

Si estás interesado en el proyecto y te gustaría sumar fuerzas para que siga creciendo y mejorando, podés abrir un hilo de discusión para charlar tus propuestas en [este link](https://github.com/gotoiot/app-fullstack-base/issues/new). Así mismo podés leer el archivo [Contribuir.md](https://github.com/gotoiot/gotoiot-doc/wiki/Contribuir) de nuestra Wiki donde están bien explicados los pasos para que puedas enviarnos pull requests.

## Sobre Goto IoT 📖

Goto IoT es una plataforma que publica material y proyectos de código abierto bien documentados junto a una comunidad libre que colabora y promueve el conocimiento sobre IoT entre sus miembros. Acá podés ver los links más importantes:

* **[Sitio web](https://www.gotoiot.com/):** Donde se publican los artículos y proyectos sobre IoT. 
* **[Github de Goto IoT:](https://github.com/gotoiot)** Donde están alojados los proyectos para descargar y utilizar. 
* **[Comunidad de Goto IoT:](https://groups.google.com/g/gotoiot)** Donde los miembros de la comunidad intercambian información e ideas, realizan consultas, solucionan problemas y comparten novedades.
* **[Twitter de Goto IoT:](https://twitter.com/gotoiot)** Donde se publican las novedades del sitio y temas relacionados con IoT.
* **[Wiki de Goto IoT:](https://github.com/gotoiot/doc/wiki)** Donde hay información de desarrollo complementaria para ampliar el contexto.

## Muestas de agradecimiento 🎁

Si te gustó este proyecto y quisieras apoyarlo, cualquiera de estas acciones estaría más que bien para nosotros:

* Apoyar este proyecto con una ⭐ en Github para llegar a más personas.
* Sumarte a [nuestra comunidad](https://groups.google.com/g/gotoiot) abierta y dejar un feedback sobre qué te pareció el proyecto.
* [Seguirnos en twitter](https://github.com/gotoiot/doc/wiki) y dejar algún comentario o like.
* Compartir este proyecto con otras personas.

## Autores 👥

Las colaboraciones principales fueron realizadas por:

* **[Agustin Bassi](https://github.com/agustinBassi)**: Ideación, puesta en marcha y mantenimiento del proyecto.
* **[Ernesto Giggliotti](https://github.com/ernesto-g)**: Creación inicial del frontend, elección de Material Design.
* **[Brian Ducca](https://github.com/brianducca)**: Ayuda para conectar el backend a la base de datos, puesta a punto de imagen de Docker.

También podés mirar todas las personas que han participado en la [lista completa de contribuyentes](https://github.com/###/contributors).

## Licencia 📄

Este proyecto está bajo Licencia ([MIT](https://choosealicense.com/licenses/mit/)). Podés ver el archivo [LICENSE.md](LICENSE.md) para más detalles sobre el uso de este material.

---

**Copyright © Goto IoT 2021** ⌨️ [**Website**](https://www.gotoiot.com) ⌨️ [**Group**](https://groups.google.com/g/gotoiot) ⌨️ [**Github**](https://www.github.com/gotoiot) ⌨️ [**Twitter**](https://www.twitter.com/gotoiot) ⌨️ [**Wiki**](https://github.com/gotoiot/doc/wiki)


CHANGELOG.md
# DAW Base App - Changes Log

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/) and this project adheres to [Semantic Versioning](http://semver.org/).


## 3.0.2
* Project modification
    * Modifica docker-compose.yml para instalar dependencias de backend antes de arrancar


## 3.0.1
* Project modification
    * Se agrega en el README.md la seccion "Lanzar la aplicacion" donde se explica como lanzar la aplicacion


## 3.0.0

* Project modification
    * Implementa para el backend el soporte para recuperar todos los dispositivos de la base de datos
    * Implementa para el backend el soporte para recuperar un dispositivo especifico de la base de datos
    * Implementa para el backend el soporte crear un nuevo dispositivo en la base de datos
    * Implementa para el backend el soporte eliminar un dispositivo existente de la base de datos
    * Implementa para el backend el soporte editar un dispositivo existente en la base de datos
    * Implementa para el front la logica para listar los dipositivos devueltos por el backend
    * Implementa para el front la logica para crear un nuevo dispositivo en el backend
    * Implementa para el front la logica para editar un dispositivo existente en el backend
    * Implementa para el front la logica para eliminar un dispositivo existente en el backend


## 2.2.0

* Project modification
    * Adds TypeScript compiler service to Docker Compose
    * Reestructures frontend folder for TypeScript
    * Adds new info to README accordingly
    * Changes project architecture image

## 2.1.0

* Project modification
    * Enhaces README accordingly to Goto IoT
    * Adds example of finished application
    * Removes unnecessary frontend images
    * Changes src code folders names

## 2.0.0

* Project modification
    * Changes project and organization names
    * Removes Typescript container
    * Removes Typescript Code
    * Executes Javascript code directly
    * Changes licence to MIT
    * Modifies README accordingly

## 1.0.0

* Project creation
    * Docker Compose implementation for whole project.
    * Typescript compilation into docker-compose.
    * MySQL 5.7 DB Server.
    * PHPMyAdmin.
    * NodeJS backend application.
    * Materialize CSS framework.


src/frontend/ts/common.ts

/*------------------------------------------------------------------*
 * common.ts														*
 *																	*
 * Historico:	07/12/2021	Primer version							*
 *																	*
 * Definiciones comunes                                             *
 *------------------------------------------------------------------*/

 // Nombre de las divs que se utilizan para mostrar los diferentes elementos
 // involucrados
 const DIV_TITULO     = 'titulo';
 const DIV_PRIMARIA   = 'primaria';
 const DIV_SECUNDARIA = 'secundaria';
 const DIV_TERCIARIA  = 'terciaria';

 // Pagina principal
 const PAGE_TITLE = 'Smart Home Web Client';

 // Pagina-2
 // Identificadores asignados a los elementos del formulario (luego con esto se pueden recuperar)
const NEW_DEVICE_FORM_NAME_ID  = 'deviceName';
const NEW_DEVICE_FORM_DESC_ID  = 'deviceDesc';
const NEW_DEVICE_FORM_TYPE_ID  = 'deviceType';
const NEW_DEVICE_FORM_RANGE_ID = 'valueRange';
const NEW_DEVICE_VALUE_ID      = 'value';

// Eventos
const EVENT_ON_CLICK  = "onClick"
const EVENT_ON_CHANGE = "onChange"

const BACKEND_URL = 'http://192.168.0.2:8000';

// Definiciones de callbacks
type LIST_DEVICES_CALLBACK = (data : Device []) => void;
type ONCHANGE_CALLBACK     = (e : HTMLElement) => void;
type GET_CALLBACK          = (data : string) => void;         
type DELETE_CALLBACK       = (success : boolean) => void;
type CREATE_CALLBACK       = (data : string) => void;
type PATCH_CALLBACK        = (data : boolean) => void;


// Tipos de dispositivos
const DEVICE_TYPE_TV     = 0;
const DEVICE_TYPE_AC     = 1;
const DEVICE_TYPE_LAMP   = 2;
const DEVICE_TYPE_BLIND  = 3;

// Usado para validaciones de los largos de los campos de texto
const NAME_MAX_LEN = 32;   // usado para validar el nombre de dispositivo escogido
const NAME_MIN_LEN = 1;    // usado para validar el nombre de dispositivo escogido
const DESC_MAX_LEN = 128; // usado para validar el nombre de dispositivo escogido
const DESC_MIN_LEN = 1;   // usado para validar el nombre de dispositivo escogido

// Nombres de los iconos de materialiaze para aplicar en cada caso en la pagina-1
const DEVICE_ICON_TV    = 'tv';
const DEVICE_ICON_AC    = 'ac_unit';
const DEVICE_ICON_LAMP  = 'lightbulb_outline';
const DEVICE_ICON_BLIND = 'dehaze';

const DEVICE_TEXT_TV    = 'Televisor';
const DEVICE_TEXT_AC    = 'Aire Acondicionado';
const DEVICE_TEXT_LAMP  = 'Lampara';
const DEVICE_TEXT_BLIND = 'Cortina';

// Nombres de los botones
const BOTON_AGREGAR_NAME  = '+Agregar'; // Boton para agregar un nuevo dispositivo (pagina 1)
const BOTON_CREAR_NAME    = "Crear";    // Boton para solicitar la creacion de un nuevo dispositivo (pagina 2)
const BOTON_CANCELAR_NAME = "Cancelar"; // Boton para cancelar la edicion/creacion de un dispositivo (pagina 2)
const BOTON_APLICAR_NAME  = "Aplicar";  // Boton para aplicar los cambios sobre un dispositivo existente (pagina 2)

// Arreglos para extraer informacion de los tipos de dispositivos
const ONOF_TEMPLATES = 
[
    { 
        type : DEVICE_TYPE_TV, 
        icon : DEVICE_ICON_TV, 
        text : DEVICE_TEXT_TV,
        onStr : 'encendido',
        offStr : 'apagado'
        
    },
    {   
        type : DEVICE_TYPE_AC, 
        icon : DEVICE_ICON_AC, 
        text : DEVICE_TEXT_AC,
        onStr : 'encendido',
        offStr : 'apagado'
    },
    {   type : DEVICE_TYPE_LAMP, 
        icon : DEVICE_ICON_LAMP,
        text : DEVICE_TEXT_LAMP,
        onStr : 'encendido',
        offStr : 'apagado'
    }
];
const RANGE_TEMPLATES = 
[
    {   
        type : DEVICE_TYPE_BLIND, 
        icon : DEVICE_ICON_BLIND, 
        text : DEVICE_TEXT_BLIND,
        range : 1,
        min : 0,
        max : 100,
        minStr : 'abierta',
        maxStr : 'cerrada'
    }, 
]
const DEVICES_ICONS = 
[
    { 
        type : DEVICE_TYPE_TV, 
        icon : DEVICE_ICON_TV, 
        text : DEVICE_TEXT_TV,
    },
    {   
        type : DEVICE_TYPE_AC, 
        icon : DEVICE_ICON_AC, 
        text : DEVICE_TEXT_AC,
    },
    {   type : DEVICE_TYPE_LAMP, 
        icon : DEVICE_ICON_LAMP,
        text : DEVICE_TEXT_LAMP,
    },
    {   type : DEVICE_TYPE_BLIND, 
        icon : DEVICE_ICON_BLIND, 
        text : DEVICE_TEXT_BLIND,
    },
]




